test/infrastructure/container/docker.go

```diff
@@ -0,0 +1,359 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package container provides an interface for interacting with Docker and potentially
+// other container runtimes.
+package container
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	dockerContainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+type docker struct {
+	dockerClient *client.Client
+}
+
+// NewDockerClient gets a client for interacting with a Docker container runtime.
+func NewDockerClient() (RuntimeInterface, error) {
+	dockerClient, err := getDockerClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to created docker runtime client")
+	}
+	return &docker{
+		dockerClient: dockerClient,
+	}, nil
+}
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, fmt.Errorf("create Docker client: %v", err)
+	}
+
+	return dockerClient, nil
+}
+
+// SaveContainerImage saves a Docker image to the file specified by dest.
+func (d *docker) SaveContainerImage(ctx context.Context, image, dest string) error {
+	reader, err := d.dockerClient.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
+
+	_, err = io.Copy(tar, reader)
+	if err != nil {
+		return err
```

Comment:
I suppose a bit more error context would be nice here (something about what we're currently trying to do)

It probably makes sense to go through all `return err`'s of this file add some more error context. This will really be helpful when debugging :)

Reply:
Makes sense. I've added additional context to all error returns. That should make it a little better when trying to figure out what happened when things go wrong.